doc/functions.md

```diff
@@ -69,7 +69,8 @@ If we were able to partially define `loop?`, we could prove `False` with it.
 
 A lambda expression is an unnamed function.
 You define lambda expressions by using the `fun` keyword. A lambda expression resembles a function definition, except that instead of the `:=` token,
-the `=>` token is used to separate the argument list from the function body. As in a regular function definition,
+the `=>` token is used to separate the argument list from the function body and you cannot specify an explicit return type, it is always inferred.
```

Comment:
This is a bit misleading. We have type ascriptions. Example:
```lean
#check fun x : Nat => (x : Int)
-- Nat -> Int
```

Reply:
Ok, wow, I've never seen that one before, it is extremely weird to say "x : Int" it sounds like you are saying x is an Int which directly contradicts the earlier statement that x is a Nat.  I'm talking about this error:

![image](https://user-images.githubusercontent.com/18707114/189987535-2c93040a-7b5a-421e-88bf-49e228e90618.png)

Reply:
That's just syntactically invalid, there is no slot to put a type there in the syntax. I would be concerned about syntactic ambiguity if you could.

Reply:
I know, and I'm not proposing we add it, I'm proposing we document this very thing "there is no slot to put a type there in the syntax."

Reply:
> Ok, wow, I've never seen that one before, it is extremely weird to say "x : Int" it sounds like you are saying x is an Int which directly contradicts the earlier statement that x is a Nat.

@lovettchris I strongly recommend you read "Theorem Proving in Lean 4" again. This notation is addressed at the beginning of the book. Here is the quote from the book: "One can always specify the type T of an expression e by writing (e : T). This instructs Lean's elaborator to use the value T as the type of e when trying to resolve implicit arguments."

Reply:
I think you misunderstood my comment.  I'm not saying I've never seen "name colon type", I've just never seen type ascription used this particular way:
```lean
#check fun x : Nat => (x : Int)
```

What is => pointing at here?  It doesn't look like a function body.  What do the parens mean here?  Is that a tuple ?  No it's not a tuple.  The "x: Int" is clear, it is saying x is type Int.  But that contradicts the earlier statement "x : Nat" which says x is a Nat.  So I'm still very confused...

Reply:
Perhaps this is a "casting operation", or a type of "explicit coercion" where you are "forcing x to become an Int" -- in C++ I would have written `(Int)x`.  Is that what this is?  Lean has "casting" ? Where is casting documented?  Perhaps your term for casting is "type ascription", but that term doesn't mean "casting" to me.  I see this also works:

```lean
def natToInt (p : Nat) : Int := (p : Int)
```

But this does not work:
```lean
def natToFloat (p : Nat) : Float := (p : Float)
/-
type mismatch
  p
has type
  Nat : Type
but is expected to have type
  Float : Type
-/
```

Reply:
>  I'm not saying I've never seen "name colon type", I've just never seen type ascription used this particular way:

@lovettchris Type ascription is explained in "Theorem Proving in Lean 4," I quoted the exact passage that explains it. Chapter 2 has more information. I understand you think this book is not very accessible, but I am sure it will be much easier to understand the book now that you have more experience with Lean.

Reply:
where exactly?

![image](https://user-images.githubusercontent.com/18707114/189998690-2129b117-053f-4e76-b2ce-861255c8e807.png)

Reply:
@lovettchris Did you read my message? I gave the exact quote.

Reply:
If you are talking about the paragraph starting with `One can always specify the type T of an expression e by writing (e : T). This instructs Lean's elaborator to use the value T as the type of e when trying to resolve implicit arguments` then sure, this is the general definition of `e : T`, but it says nothing here about using that as a method body in a type of very limited type coercion operation...

Reply:
@lovettchris This is not the general definition of `e : T`. Please read the book again. It is introducing the notation here, but it is not the only occurrence. The notion of coercion is introduced later. The type ascription `(e : T)` is used to specify the expected type. Everything will make more sense after you read the book again.

Reply:
Unless Lean has some other general behavior that says any expression can be used anywhere, so of course you can ascribe `(e : T)` as the entirety of a function body... or something.  In playing with this I see:

```lean
def NatToInt (p: Nat) := 
  let x := p
  (x : Int)
```
works too, but x is not Nat in `let x := p`, so this means `(x : Int)` is indeed a coercion.  It's saying the expression "x" is the return value, but I want you to change the type of this expression to `Int` using the special syntax `(x : Nat)` which requires parentheses.  Perhaps this is what the Theorem Proving paragraph was trying to say, but I didn't read it that way, I read it as just explaining the generally "name colon type" syntax that is used everywhere in Lean...

Reply:
@lovettchris If you want to help me, please read the book. It will save us time, and you will have a much better understanding of what is going on.

Reply:
I see now it works more generally too:
```lean4
def NatToIntPair : Nat × Nat →  Int × Int 
  | (x, y) => ((x : Int), (y : Int))
```
so it's starting to make sense.  Perhaps this is just a case of "the TPIL book is very dense" and I missed nuance here that you can use this type ascription syntax to force some simple type coercions to happen, a bit like casting, but cannot cast Float to Int or Int to Float or Int to Nat, it can only do Nat to Int...

Reply:
> Perhaps this is just a case of "the TPIL book is very dense" and I missed nuance here that you can use this type ascription syntax to force some simple type coercions to happen, a bit like casting, but cannot cast Float to Int or Int to Float or Int to Nat, it can only do Nat to Int...

TPIL will make much more sense in a second read. TPIL has exercises, I suggest you try to solve all of them. It will help for use. Many people in the community managed to learn Lean this way and succeeded.

Reply:
Well, I recompiled Lean without `instance : Coe Nat Int := ⟨Int.ofNat⟩` and indeed the `#check fun x : Nat => (x : Int)` no longer compiles, so (e : T) is not only specifying the type of the expression it is triggering coercion as described in [Coercions using Type Classes](https://leanprover.github.io/theorem_proving_in_lean4/type_classes.html#coercions-using-type-classes).  The problem with that section is it glosses over this easy case never showing this example, and dives right into much more complicated stuff involving _class of sorts_ and _class of function types_ and so on, and not once in the entire section does it use `(e : T)` syntax as a forced coercion.  It is shown in an #eval in https://leanprover.github.io/lean4/doc/nat.html but that is making the literal 5 be interpreted as an Int, and this is not coercion because that still compiles after I made that change to Lean. PS: There are also no example exercises under `Coercions using Type Classes`... 

But I now understand that (e : T) works for any coercion, for example, I see there's an `instance : Coe String IO.Error := ⟨IO.userError⟩` and so I can write `#check ("test" : IO.Error)` and get an `-- IO.userError "test" : IO.Error`.  Previously when reading the Coercions section and reading lots of Lean code I had only gleaned that Coercions allowed you to do this sort of thing where coercion is transparently happening to match specified arg types or return types.  

```lean
def foo (e : IO.Error) :=  e
#eval foo "test"    
```
In fact, in my change to Lean removing `instance : Coe Nat Int := ⟨Int.ofNat⟩` I found only 3 instances of `(e : Int)` in the entire code base, (2 in JSon Basic.lean and 1 in Congr.lean) so I think it's not too surprising I somehow missed it.

Reply:
So to close this thread, I removed the contentious statement about no declared return type and added this paragraph which adds the missing information and link between type ascription and coercion:

> While you cannot specify an explicit return type in a lambda expression you can use the type ascription
expression `(e : T)` in the body to declare a new type for the expression `e`.
For example: `λ (x : Nat) => (x : Int)` is able to convert a given value of type `Nat` to an `Int` using coercion.